libnetwork/drivers/bridge/setup_ip_forwarding.go

```diff
@@ -33,42 +27,93 @@ func setupIPForwarding(enableIPTables bool, enableIP6Tables bool) error {
-		// drop only if the daemon option iptables is not set to false.
-		if enableIPTables {
-			iptable := iptables.GetIptable(iptables.IPv4)
-			if err := iptable.SetDefaultPolicy(iptables.Filter, "FORWARD", iptables.Drop); err != nil {
-				if err := configureIPForwarding(false); err != nil {
-					log.G(context.TODO()).Errorf("Disabling IP forwarding failed, %v", err)
+		defer func() {
+			if retErr != nil {
+				if err := os.WriteFile(ipv4ForwardConf, []byte{'0', '\n'}, forwardConfPerm); err != nil {
+					log.G(context.TODO()).WithError(err).Error("Disabling IPv4 forwarding failed")
 				}
+			}
+		}()
+
+		// When enabling ip_forward set the default policy on forward chain to drop.
+		if wantFilterForwardDrop {
+			if err := setFilterForwardDrop(iptables.IPv4); err != nil {
 				return err
 			}
-			iptables.OnReloaded(func() {
-				log.G(context.TODO()).Debug("Setting the default DROP policy on firewall reload")
-				if err := iptable.SetDefaultPolicy(iptables.Filter, "FORWARD", iptables.Drop); err != nil {
-					log.G(context.TODO()).Warnf("Setting the default DROP policy on firewall reload failed, %v", err)
-				}
-			})
 		}
 	}
+	return nil
+}
 
-	// add only iptables rules - forwarding is handled by setupIPv6Forwarding in setup_ipv6
-	if enableIP6Tables {
-		iptable := iptables.GetIptable(iptables.IPv6)
-		if err := iptable.SetDefaultPolicy(iptables.Filter, "FORWARD", iptables.Drop); err != nil {
-			log.G(context.TODO()).Warnf("Setting the default DROP policy on firewall reload failed, %v", err)
+func setupIPv6Forwarding(wantFilterForwardDrop bool) (retErr error) {
+	configUpdated := false
+
+	// Get current IPv6 default forwarding setup
+	// FIXME(robmry) - is it necessary to set this, setting "all" (below) does the job?
+	ipv6ForwardDataDefault, err := os.ReadFile(ipv6ForwardConfDefault)
+	if err != nil {
+		return fmt.Errorf("Cannot read IPv6 default forwarding setup: %v", err)
+	}
+	// Enable IPv6 default forwarding only if it is not already enabled
+	if ipv6ForwardDataDefault[0] != '1' {
+		if err := os.WriteFile(ipv6ForwardConfDefault, []byte{'1', '\n'}, forwardConfPerm); err != nil {
+			log.G(context.TODO()).Warnf("Unable to enable IPv6 default forwarding: %v", err)
 		}
-		iptables.OnReloaded(func() {
-			log.G(context.TODO()).Debug("Setting the default DROP policy on firewall reload")
-			if err := iptable.SetDefaultPolicy(iptables.Filter, "FORWARD", iptables.Drop); err != nil {
-				log.G(context.TODO()).Warnf("Setting the default DROP policy on firewall reload failed, %v", err)
+		configUpdated = true
+	}
+	defer func() {
+		if retErr != nil {
+			if err := os.WriteFile(ipv6ForwardConfDefault, []byte{'0', '\n'}, forwardConfPerm); err != nil {
+				log.G(context.TODO()).WithError(err).Error("Disabling IPv6 default forwarding failed")
 			}
-		})
+		}
+	}()
+
+	// Get current IPv6 all forwarding setup
+	ipv6ForwardDataAll, err := os.ReadFile(ipv6ForwardConfAll)
```

Comment:
Maybe it'd be benefitial here too to read both files, early-return if both are already set, and if not then start with ipt default policy and then set both unconditionnally. That'd make the code simpler I think.

Reply:
As above - we'd still need to undo anything done so-far, so I'm not sure it'd help?

But I've put the undo-sysctl code in the right place now (oops), so perhaps it makes more sense?

Reply:
It's existing code, but curious; we _fail_ this function if we fail to read, but we _don't_ return an error when failing to update it? Could that have situations where we would not actually be updating the config? (Are those save to be ignored?).

Reply:
Looks like for IPv4 we have a separate utility (`configureIPForwarding()`). Wondering if (not for this PR), we should have that utility handle the "changed" or "not changed" case, e.g. something like;

```go
func configureIPForwarding(file string, enable bool) (changed bool, _ error) {
	data, err := os.ReadFile(file)
	if err != nil {
		return false, fmt.Errorf("cannot read IP forwarding setup: %w", err)
	}
	if enabled := data[0] == '1'; enabled == enable {
		return false, nil
	}
	var val byte
	if enable {
		val = '1'
	}
	if err := os.WriteFile(file, []byte{val, '\n'}, 0o644); err != nil {
		return false, fmt.Errorf("cannot write IP forwarding setup: %w", err)
	}
	return true, nil
}
```

That would return a boolean to indicate we changed anything, but an error when failing to either read or update (but that depends a bit on my previous comment).

The above wraps the errors, so we could detect `os.IsnotFound()` situations (if those are ok to be ignored).

Reply:
Right, yes ... things won't work properly if IPv6 forwarding is needed and it can't be enabled, and I can't see why the error handling should be different to IPv4 - so I'll do that, thank you.